support-workers/src/test/scala/com/gu/support/workers/integration/GiftSubcriptionZuoraSpec.scala

```diff
@@ -0,0 +1,64 @@
+package com.gu.support.workers.integration
+
+import com.gu.salesforce.Fixtures.idId
+import com.gu.support.redemption.GetCodeStatus.{CodeAlreadyUsed, NoSuchCode}
+import com.gu.support.redemption.generator.CodeBuilder.GiftCode
+import com.gu.support.redemption.generator.GiftCodeGeneratorService
+import com.gu.support.redemptions.{RedemptionCode, RedemptionData}
+import com.gu.support.workers.JsonFixtures.{createDigiPackGiftRedemptionJson, createDigiPackGiftSubscriptionJson}
+import com.gu.support.workers.{AsyncLambdaSpec, BillingPeriod, Fixtures, MockContext, RequestInfo}
+import com.gu.support.workers.lambdas.{DigitalSubscriptionGiftRedemption, HandlerResult}
+import com.gu.support.workers.states.{CreateZuoraSubscriptionState, SendThankYouEmailState}
+import com.gu.test.tags.annotations.IntegrationTest
+import org.mockito.ArgumentMatchers.any
+import io.circe.parser.decode
+
+import scala.concurrent.Future
+
+@IntegrationTest
+class GiftSubcriptionZuoraSpec extends CreateZuoraSubscriptionBaseSpec {
+
+  val giftCode = GiftCode("gd03-00000000").get
+  val mockCodeGenerator = mock[GiftCodeGeneratorService]
+  when(mockCodeGenerator.generateCode(any[BillingPeriod])).thenReturn(giftCode)
+
+  // This spec relies on the fact that tests within a spec are executed in sequence
+  // So we can create a sub, redeem it and then try to redeem it again to get an error
+
+  "CreateZuoraSubcription" should "create a Digital Pack gift subscription" in {
+    createSubscription(createDigiPackGiftSubscriptionJson, mockCodeGenerator)
+      .map { case (_, maybeError, _) => maybeError shouldBe None }
+  }
+
+  "DigitalSubscriptionGiftRedemption" should "redeem a Digital Pack gift subscription" in
+    redeemSubscription(giftCode.value)
+      .map(_.value.user.id shouldBe idId)
+
+  it should "throw a CodeAlreadyUsed exception redeem a Digital Pack gift subscription" in
+    recoverToExceptionIf[RuntimeException](
+      redeemSubscription(giftCode.value)
+    ).map(_.getMessage shouldBe CodeAlreadyUsed.clientCode)
+
+
+  it should "throw a NoSuchCode exception redeem a Digital Pack gift subscription" in {
```

Comment:
for even better checking, move this test to the top, and use the generated code.  Then we have proved the selected code does not exist before we create it.

Reply:
That is a good idea. It will only work if we generate a new code every test run rather than using the same one every time but I think that's fine.

Reply:
oh yeah didn't spot that, thanks for updating it 👍